staticjinja/staticjinja.py

```diff
@@ -93,39 +117,40 @@ def get_template(self, template_name):
         """
         return self._env.get_template(template_name)
 
-    def _get_context_generator(self, template_name):
-        """Get a context generator for a template.
-
-        Raises a :exc:`ValueError` if no matching context generator can be
-        found.
-
-        :param template_name: the name of the template
-        """
-        for regex, context_generator in self.contexts:
-            if re.match(regex, template_name):
-                return context_generator
-        raise ValueError("no matching context generator")
-
     def get_context(self, template):
         """Get the context for a template.
 
         By default, this function will return an empty context.
 
-        If a matching context_generator can be found, it will be passed the
-        template and executed. The context generator should return a dictionary
-        representing the context.
+        If a matching context_generator can be found and is a function
+        it will be executed after being passed the template if it expects an
+        argument. If it is not a function then it will be assumed to act like
+        a dictionary object.
+
+        If several context_generator can be found, their resulting dictionary
+        will be merged if mergecontexts is True or only the first one will be
+        used otherwise.
 
         :param template: the template to get the context for
         """
-        try:
-            context_generator = self._get_context_generator(template.name)
-        except ValueError:
-            return {}
-        else:
-            try:
-                return context_generator(template)
-            except TypeError:
-                return context_generator()
+
+        context = {}
+        for regex, context_generator in self.contexts:
+            if re.match(regex, template.name):
+                if inspect.isfunction(context_generator):
+                    if _has_argument(context_generator):
+                        context.update(context_generator(template))
+                    else:
+                        context.update(context_generator())
```

Comment:
I rather strongly feel the current code is pretty close to being a bug. A user who writes a context_generator function which accidentally raises a TypeError will have a hard time understanding what happens next without reading staticjinja's source.

I also think my version is easier to read, especialy since it doesn't require reading the docstring of the preceding function to understand how the exception is meant to arise. But I'm clearly biaised since I wrote my version...

I do see I'm guilty of some duck typing violation here but I don't think it prevents anyone from doing anything. Note I never check anything is a dict. By the way, this is why I wrote somewhere "acts as an dict" and not "is a dict" but I I agree it's confusing (I hesitated to write that it could be an OrderedDict for instance because I didn't actually check it would be ok). I take the opportunity to mention that I don't comment on your comments about documentation because I 100% agree your modifications are better.

Anyway this would be only about code style (on which I obviously don't claim any authority) without this accidental catch issue which is my main point here.